0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.
+Just producing a sequence of the values in a set is simple using a sequence comprehension expression: `seq i | i in s`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## foreach Loops
+
+A `foreach` loop in Dafny resembles `for` or `foreach` loops in many other mainstream languages,
+but is substantially more general and expressive. A very simple example only looks slightly different than expected:
+
+```dafny
+foreach x | x in [1, 2, 3, 4, 5] {
+  print x, "\n";
+}
+```
+
+The feature supports much more sophisticated uses as well, however,
+including binding multiple variables at once and filtering:
+
+```dafny
+var myDoubleDeckerSeq: seq<seq<int>> := ...;
+foreach x, y | x in myDoubleDeckerSeq && y in x && y != 0 {
+  Process(y);
+}
+```
+
+The high-level syntax for a `foreach` loop reuses several existing concepts:
+
+```
+ForeachLoop ::=
+  "foreach"
+  QuantifierDomain
+  LoopSpec
+  Body
+```
+
+More specifically, this expands to:
+
+```
+ForeachLoop ::=
+  "foreach" 
+  Ident [":" Type] { "," Ident [":" Type] } [ "|" Expression ]
+  { InvariantDecl | ModifiesDecl | DecreasesDecl }
+  Body
+```
+
+A `foreach` loop closely resembles a `forall` statement, the key difference being that a `forall`
+loop executes its body once for every tuple of quantified values simultaneously, whereas a `foreach` loop
+executes its body once for each tuple of quantified values in sequence, one at a time.
+
+Similarly to set comprehensions or assign-such-that statements, the domain of a `foreach` loop
+must be enumerable. The following loop would produce the error `"the domain of a foreach loop must be enumerable,
+but Dafny's heuristics can't figure out how to produce an enumeration for 'x'"`.
+
+```dafny
+foreach x: real | 0.0 <= x < 1.0 {
+  ...
+}
+```
+
+The domain is allowed to be potentially infinite if the loop is used in a compiled context and an explicit `decreases` clause is provided.
+`decreases *` is permitted, in which the `foreach` loop may never terminate. Any other `decreases` clause can be provided
+to ensure the loop terminates even if the domain is potentially infinite. The following (slightly bizarre) example is legal:
+
+```dafny
+foreach x: nat 
+  decreases 10 - x
+{
+  print x, "\n";
+  if x == 10 { break; }
+}
+```
+
+Most `foreach` loops will take the form `foreach x | x in c { ... }`, where `c` is a `seq`, `set` or `multiset`.
+This includes expressions like `m.Keys`, `m.Values` and `m.Items` when `m` is a `map`.
+Looping over the keys and values of a map is particularly readable and aesthetically pleasing with this syntax,
+which also makes it easier for the compiled program to avoid unnecessary intermediate tuple values:
+
+```dafny
+foreach k, v | k in m.Keys && m[k] == v {
+  print "m[", k, "] == ", v;
+}
+```
```

Comment:
Oh, I like this!

Reply:
I liked it even more when I just had

```dafny
foreach k, v | m[k] == v {
  print "m[", k, "] == ", v;
}
```

but then realized I needed the `k in m.Keys` part. :)